tests/Tests/ORM/Hydration/ComplexObjectHydratorTest.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Hydration;
+
+use Doctrine\ORM\AbstractQuery;
+use Doctrine\ORM\Mapping\Driver\AttributeDriver;
+use Doctrine\ORM\Tools\SchemaTool;
+use Doctrine\Tests\Models\ComplexHydration\Item;
+use Doctrine\Tests\Models\ComplexHydration\ItemSerie;
+use Doctrine\Tests\Models\ComplexHydration\Serie;
+use Doctrine\Tests\Models\ComplexHydration\SerieImportator;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+use function count;
+use function dirname;
+use function sprintf;
+use function substr;
+
+/**
+ * @requires PHP 8.1
+ */
+class ComplexObjectHydratorTest extends OrmFunctionalTestCase
+{
+    /** @var bool */
+    private $entitiesCreated = false;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->_em         = $this->getEntityManager(null, new AttributeDriver([dirname(__DIR__, 2) . '/Models/ComplexHydration'], true));
+        $this->_schemaTool = new SchemaTool($this->_em);
+
+        if (! $this->entitiesCreated) {
+            $this->entitiesCreated = true;
+
+            $this->setUpEntitySchema([Serie::class, Item::class, SerieImportator::class, ItemSerie::class]);
+
+            for ($iSerie = 1; $iSerie <= 2; $iSerie++) {
+                $serie = new Serie();
+                $serie->setLibelle('Serie ' . $iSerie);
+                $this->_em->persist($serie);
+
+                $serieImportator = new SerieImportator();
+                $serieImportator->setLibelle('imp ' . $iSerie);
+                $serieImportator->setSerie($serie);
+                $this->_em->persist($serieImportator);
+
+                for ($iItem = 1; $iItem <= 12; $iItem++) {
+                    $item = new Item();
+                    $item->setLibelle('Item ' . $iSerie . '.' . $iItem);
+
+                    $itemSerie = new ItemSerie();
+                    $itemSerie->setSerie($serie);
+                    $itemSerie->setItem($item);
+                    $itemSerie->setNumber('Num ' . $iSerie . '.' . $iItem);
+
+                    $this->_em->persist($item);
+                    $this->_em->persist($itemSerie);
+                }
+            }
+
+            $this->_em->flush();
+            $this->_em->clear();
+        }
+    }
+
+    protected function tearDown(): void
+    {
+        $this->_em->createQuery('DELETE FROM ' . ItemSerie::class)->execute();
+        $this->_em->createQuery('DELETE FROM ' . SerieImportator::class)->execute();
+        $this->_em->createQuery('DELETE FROM ' . Serie::class)->execute();
+        $this->_em->createQuery('DELETE FROM ' . Item::class)->execute();
+    }
+
+    protected function getDQL(): string
+    {
+        return sprintf(
+            <<<'SQL'
+               SELECT s0_, i1_, i2_, s4_, s5_, i6_
+                 FROM %s s0_
+            LEFT JOIN s0_.itemSeries i1_
+            LEFT JOIN i1_.item i2_
+            LEFT JOIN s0_.serieImportators s4_
+            LEFT JOIN s4_.serie s5_
```

Comment:
s0 is a serie, and then you join on serieImportators, and then… back on serie?!? This does not make much sense, does it?

Reply:
Since you reported an issue, you must have a realistic scenario that you tried to abstract away… but if it's so abstract that nobody understands the abstraction, it's not great. Tests should be simple to understand.

Reply:
I agree with you, i had a automatic DQL builder and I had this but it's not useful.
But the bug apear in this case, only if there a 11+ itemSeries (Why?) and always the second item is second to last I don't really understand why

Reply:
I don't know what responds, it's a bug, the correction is "just" change key() to indexof($element), 
I understand it take the last element instead of the one whitch correspond to the actual key,
It's only the tests I wrote there are difficult to understand, I just think the bug can appear in other cases so I try to repair it, now if you think it"s not helpful I can hear it,
even if I find that it remains a bug.

Reply:
You say it's a bug, but what your test shows me is that it might actually be a performance optimization built on the assumption that nobody is going to write something as stupid as `author a LEFT JOIN a.books b LEFT JOIN b.author a2`… I mean, what is the point of that?

Reply:
it does not take the entity corresponding to the foreign key, Ok with a studip query but it's more a bug than a performance optimisation.
Take a look to my reproducer it's maybe more visual ; https://github.com/eltharin/errordoctrine/
the visual result : 
![303400982-0967d501-06c9-4bee-bbc7-d549f4109ea6](https://github.com/doctrine/orm/assets/7547802/cd491cd2-7c66-4236-9c17-691b532a03df)
As you can see, the last itemSeries element does'nt take item X.12 but X.2 but the foreign key is pointing to X.12